downstairs/src/region.rs

```diff
@@ -876,37 +881,54 @@ impl Extent {
             // contiguous with the request before it. Since we're counting
             // pairs, and the number of pairs is one less than the number of
             // requests, we need to add 1 to get our actual run length.
-            let n_contiguous_requests = requests[req_run_start..]
-                .windows(2)
-                .take_while(|reqs| {
-                    reqs[0].offset.value + 1 == reqs[1].offset.value
-                })
-                .count()
-                + 1;
+            let mut n_contiguous_requests = 1;
+
+            for request_window in requests[req_run_start..].windows(2) {
+                if (request_window[0].offset.value + 1
+                    == request_window[1].offset.value)
+                    && ((n_contiguous_requests + 1) < self.iov_max)
+                {
+                    n_contiguous_requests += 1;
+                } else {
+                    break;
+                }
+            }
 
             // Create our responses and push them into the output. While we're
-            // at it, check for overflows
+            // at it, check for overflows. Create iovecs alongside the
+            // responses.
             let resp_run_start = responses.len();
+            let mut iovecs = Vec::with_capacity(n_contiguous_requests);
             for req in requests[req_run_start..][..n_contiguous_requests].iter()
             {
                 let resp =
                     ReadResponse::from_request(req, self.block_size as usize);
                 self.check_input(req.offset, &resp.data)?;
+                iovecs.push(libc::iovec {
+                    iov_base: resp.data.as_ptr() as *mut libc::c_void,
+                    iov_len: resp.data.len(),
+                });
                 responses.push(resp);
             }
 
             // Finally we get to read the actual data. That's why we're here
             cdt::extent__read__file__start!(|| {
                 (job_id, self.number, n_contiguous_requests as u64)
             });
-            inner.file.seek(SeekFrom::Start(
-                first_req.offset.value * self.block_size,
-            ))?;
-            let mut read_buffer = BytesMut::with_capacity(
-                n_contiguous_requests * self.block_size as usize,
-            );
-            read_buffer.resize(read_buffer.capacity(), 0);
-            inner.file.read_exact(&mut read_buffer)?;
+
+            let bytes_read = unsafe {
```

Comment:
Do we have a good understanding of why the `unsafe` is needed here? 
It makes me nervous, but only because I don't understand it.

Reply:
I don't understand unsafe either :) I think if all we're doing is calling this function in a very small unsafe block and not doing any memory chicanery, we're ok.